lib/backend/helpers.go

```diff
@@ -61,3 +62,36 @@ func ReleaseLock(ctx context.Context, backend Backend, lockName string) error {
 	}
 	return nil
 }
+
+// jitter is used to allow multi-process reattempts to be randomly dispersed.
+var jitter = utils.NewJitter()
+
+// Swap emulates an upsert operation in that it will update an existing item,
+// otherwise create the item. Performs an additional operation to check if user exists first.
+// This allows us to differentiate between an update/create operation.
+func Swap(ctx context.Context, backend Backend, item Item) (*Item, error) {
```

Comment:
I would not fix the issue with `upsert` pattern with `Swap` way . Instead I would gradually start replacing `upserts` in our code with a proper create/update method. For example, during social login, I would check if this user exists before trying to create one (btw, does it need to be updated if it already exists?).  
`Swap` may look as a good temporary workaround but it actually further postpones us from fixing the problem.

Reply:
I am okay with going with either approach. There wasn't really a problem with the Update/Create approach, just that these two operations were nearly identical and it introduced extra branching where every time we must decide we want to create or update, we must check if user exists. With UpsertUser with swap it removed that necessary branching and did the checking under hood.

Examples of code snippets:
- [the branching with Create/Update approach](https://github.com/gravitational/teleport/blob/dc1d1ff649771382c7cffe4b988b970be366da01/lib/auth/github.go#L381)
- [the UpsertUser swap approach](https://github.com/gravitational/teleport/blob/d553836e5ccbb2a6bd6e8c841a3adbde3464beae/lib/auth/github.go#L387)

I guess another problem with UpsertUser with swap, is if you are checking if a user existed prior to calling UpsertUser, get query gets called twice...

Reply:
@alex-kovoy and @awly reasoning made sense, so I reverted back to Create/Update approach:
- more flexible (avoids double get queries, b/c most times it seems we have to check if user existed prior to calling Upsert)
- explicit operations
- simple/predictable behavior